mapper/mapper.go

```diff
@@ -1,203 +1,243 @@
 package mapper
 
 import (
-	"context"
-	"net/url"
-	"strconv"
-	"strings"
-
 	searchC "github.com/ONSdigital/dp-api-clients-go/site-search"
 	model "github.com/ONSdigital/dp-frontend-models/model/search"
 	"github.com/ONSdigital/dp-frontend-search-controller/data"
-	"github.com/ONSdigital/log.go/log"
 )
 
-// CreateSearchPage maps type cookies.Policy to model.Page
-func CreateSearchPage(ctx context.Context, query url.Values, respC searchC.Response, categories []data.Category) (page model.Page) {
+// CreateSearchPage maps type searchC.Response to model.Page
+func CreateSearchPage(validatedQueryParams data.SearchURLParams, categories []data.Category, respC searchC.Response) (page model.Page) {
 	// SEARCH STRUCT MAPPING
-	var err error
 	page.Metadata.Title = "Search"
 	page.SearchDisabled = true
-	page.Data.Query = query.Get("q")
 
-	page.Data.Filter.Query = query["filter"]
-	page.Data.Filter.Options = []string{"Publication", "Data", "Other"}
+	mapQuery(&page, validatedQueryParams, categories, respC)
 
-	page.Data.Sort.Query = query.Get("sort")
-	page.Data.Sort.LocaliseFilterKeys = getFilterSortKeyList(query, categories)
-	page.Data.Sort.FilterText = strings.Replace(strings.ToLower(page.Data.Sort.Query), "-", " ", 1)
+	mapResponse(&page, respC, categories)
 
-	if query.Get("limit") != "" {
-		page.Data.Limit, err = strconv.Atoi(query.Get("limit"))
-		if err != nil {
-			log.Event(ctx, "unable to convert search limit to int - default to limit 10", log.INFO)
-			page.Data.Limit = 10
-		}
-	}
+	return page
+}
 
-	if query.Get("offset") != "" {
-		page.Data.Offset, err = strconv.Atoi(query.Get("offset"))
-		if err != nil {
-			log.Event(ctx, "unable to convert search offset to int - default to offset 0", log.INFO)
-			page.Data.Offset = 0
-		}
+func mapQuery(page *model.Page, validatedQueryParams data.SearchURLParams, categories []data.Category, respC searchC.Response) {
+	page.Data.Query = validatedQueryParams.Query
+
+	page.Data.Filter = validatedQueryParams.Filter.Query
+
+	mapSort(page, validatedQueryParams, categories)
+
+	mapPagination(page, validatedQueryParams, categories, respC)
+}
+
+func mapSort(page *model.Page, validatedQueryParams data.SearchURLParams, categories []data.Category) {
+	page.Data.Sort.Query = validatedQueryParams.Sort.Query
+
+	page.Data.Sort.LocaliseFilterKeys = validatedQueryParams.Filter.LocaliseKeyName
+
+	page.Data.Sort.LocaliseSortKey = validatedQueryParams.Sort.LocaliseKeyName
+
+	pageSortOptions := []model.SortOptions{}
```

Comment:
I think it is good practice to declare as a nil slice rather than an empty slice. This might not mean much in some situations however if it is passed to the frontend via json, then `nil` turns into `null`.
```suggestion
	var pageSortOptions []model.SortOptions
```
You can still append to the nil slice e.g.
https://tour.golang.org/moretypes/15

What impact does this have?
The frontend will be given an empty array `[]` in the JSON currently but if we declare this as suggested then if nothing is appended to the slice then the frontend will get `null`.

I am aware this is also done elsewhere so will only raise for new entries in this PR

Reply:
So you can see the impact on the frontend with JS:

![Screenshot 2021-03-11 at 11 18 18](https://user-images.githubusercontent.com/47502916/110779521-88947e80-825b-11eb-86b1-a6f14e20eef5.png)

Also worth noting this is a frontend usage example, but there could be others who use our APIs as a whole and if this lands in the API then they too will get `[]` rather than `null` once Go converts it to JSON. So always worth declaring slices as nil rather than a literal.

Reply:
That is awesome. This makes sense